Mathlib/Data/Finset/Basic.lean

```diff
@@ -4007,3 +4010,30 @@ end List
 -- Note that we cannot use `List.sublists` itself as that is defined very early.
 assert_not_exists List.sublistsLen
 assert_not_exists Multiset.powerset
+
+namespace Mathlib.Meta
+open Qq Lean Meta Finset
+
+/-- Attempt to prove that a finset is nonempty using the `finsetNonempty` aesop rule-set.
+
+You can add lemmas to the rule-set by tagging them with either:
+* `aesop safe apply (rule_sets [finsetNonempty])` if they are always a good idea to follow or
+* `aesop unsafe apply (rule_sets [finsetNonempty])` if they risk directing the search to a blind
+  alley.
+-/
+def proveFinsetNonempty {u : Level} {α : Q(Type u)} (s : Q(Finset $α)) :
+    MetaM (Option Q(Finset.Nonempty $s)) := do
+  -- Aesop expects to operate on goals, so we're going to make a new goal.
+  let goal ← Lean.Meta.mkFreshExprMVar q(Finset.Nonempty $s)
+  let mvar := goal.mvarId!
+  -- We want this to be fast, so use only the basic and `Finset.Nonempty`-specific rules.
+  let rulesets ← Aesop.Frontend.getGlobalRuleSets #[`builtin, `finsetNonempty]
+  -- Fail if the new goal is not closed.
+  let rules ← Aesop.mkLocalRuleSet rulesets { terminal := true, generateScript := false }
+  let (remainingGoals, _) ← Aesop.search mvar (.some rules)
```

Comment:
Does this fail in the monadic sense, or in the sense of hitting the  `return none` below?

Reply:
This is a monadic failure afaict. So you mean it should be:
```suggestion
  -- Fail if the new goal is not closed.
  let rules ← Aesop.mkLocalRuleSet rulesets { terminal := true, generateScript := false }
  let (remainingGoals, _) ← Aesop.search mvar (.some rules) <|> return .none
```

Reply:
I agree it's probably cleaner returning `none`

Reply:
Actually this is incorrect syntax and I'm not sure what the correct one is.

Reply:
Sorry! The following works on my machine:
```suggestion
  -- Fail if the new goal is not closed.
  let rules ← Aesop.mkLocalRuleSet rulesets { terminal := true, generateScript := false }
  let (remainingGoals, _) ←
    try Aesop.search mvar (.some rules)
    catch _ => return none
```